diff --git a/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp b/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp
--- a/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp
+++ b/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp
@@ -707,6 +707,7 @@ public:
   bool isUImm20() const { return IsUImm<20>(); }
   bool isUImm32() const { return IsUImm<32>(); }
   // RISCVAsmParser.cpp - riscv_operands - INSERTION_START
+  bool isUImm12() const { return IsUImm<12>(); }
   // RISCVAsmParser.cpp - riscv_operands - INSERTION_END

   bool isUImm8GE32() const {
diff --git a/llvm/lib/Target/RISCV/RISCVInstrInfo.td b/llvm/lib/Target/RISCV/RISCVInstrInfo.td
--- a/llvm/lib/Target/RISCV/RISCVInstrInfo.td
+++ b/llvm/lib/Target/RISCV/RISCVInstrInfo.td
@@ -316,6 +316,7 @@ def simm21_lsb0_jal : Operand<OtherVT> {
 }

 // RISCVInstrInfo.td - field_types - INSERTION_START
+def uimm12 : RISCVUImmLeafOp<12>;
 // RISCVInstrInfo.td - field_types - INSERTION_END

 def BareSymbol : AsmOperandClass {
