diff --git a/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp b/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp
--- a/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp
+++ b/llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp
@@ -725,6 +725,7 @@ public:
   bool isUImm48() const { return IsUImm<48>(); }
   bool isUImm64() const { return IsUImm<64>(); }
   // RISCVAsmParser.cpp - riscv_operands - INSERTION_START
+  bool isUImm12() const { return IsUImm<12>(); }
   // RISCVAsmParser.cpp - riscv_operands - INSERTION_END

   bool isUImm5NonZero() const {
diff --git a/llvm/lib/Target/RISCV/RISCVInstrInfo.td b/llvm/lib/Target/RISCV/RISCVInstrInfo.td
--- a/llvm/lib/Target/RISCV/RISCVInstrInfo.td
+++ b/llvm/lib/Target/RISCV/RISCVInstrInfo.td
@@ -320,6 +320,7 @@ def simm21_lsb0_jal : Operand<OtherVT> {
 }

 // RISCVInstrInfo.td - field_types - INSERTION_START
+def uimm12 : RISCVUImmLeafOp<12>;
 // RISCVInstrInfo.td - field_types - INSERTION_END

 def BareSymbol : AsmOperandClass {
